web/packages/teleterm/src/agentCleanupDaemon/agentCleanupDaemon.js

```diff
@@ -0,0 +1,142 @@
+// that can be simply forked.
+//
+// What's more, it cannot be an ESM module (.mjs). Electron doesn't support them well yet [3],
+// particularly when ESM modules are forked from asar.
+//
+// [1] https://github.com/TypeStrong/ts-node/tree/47d4f45f35e824a2515e17383a563be7dba7d8ff#native-ecmascript-modules
+// [2] https://github.com/TypeStrong/ts-node/tree/47d4f45f35e824a2515e17383a563be7dba7d8ff#how-it-works
+// [3] https://github.com/electron/electron/issues/21457
+
+const { setTimeout } = require('node:timers/promises');
+
+const agentPid = parseInt(process.argv[2], 10);
+// Pass ppid over argv rather than reading process.ppid, as ppid can change when the cleanup deamon
+// gets orphaned.
+const parentPid = parseInt(process.argv[3], 10);
+const timeToSigkill = parseInt(process.argv[4], 10) || 5_000;
+
+if (!agentPid) {
+  throw new Error('Agent PID must be passed over argv');
+}
+if (!parentPid) {
+  throw new Error('Parent PID must be passed over argv');
+}
+if (!process.send) {
+  // https://nodejs.org/docs/latest-v18.x/api/child_process.html#optionsstdio
+  // https://nodejs.org/docs/latest-v18.x/api/process.html#processsendmessage-sendhandle-options-callback
+  throw new Error(
+    'Agent cleanup daemon must be spawned from a Node.js parent that establishes an IPC channel'
+  );
+}
+
+console.log(`Spawned from PID ${parentPid}, agent PID ${agentPid}.`);
+
+// disconnect will be emitted when the IPC channel between the cleanup daemon and the parent gets
+// closed. Since we don't explicitly close the channel at any point, this means that the parent got
+// unexpectedly terminated.
+process.on('disconnect', async () => {
+  console.log(`Disconnected from the parent with PID ${parentPid}.`);
+  await terminateAgent();
+});
+
+// Signal readiness to the parent by sending an empty message.
+// In some cases, such as in tests, we need to know that the cleanup daemon is ready and operating
+// before we terminate the parent.
+process.send(null, undefined, undefined, () => {
+  // Ignore errors. In a situation where the parent got killed before the execution got to this
+  // point, process.send will fail with ERR_IPC_CHANNEL_CLOSED.
+  //
+  // We handle the IPC channel being closed below with process.connected.
+});
+
+postLaunchChecks();
+
+async function postLaunchChecks() {
+  // Terminate the agent if the parent got terminated before the cleanup daemon was able to fully
+  // set up.
+  //
+  // In that scenario, the 'disconnect' event will never be fired and the event loop will no longer
+  // have any work to perform.
+  if (!process.connected) {
+    console.error(
+      `The parent with PID ${parentPid} got terminated during setup.`
+    );
+    await terminateAgent();
+    process.exitCode = 41;
+    return;
+  }
+
+  if (!isRunning(agentPid)) {
+    console.error(
+      `The agent with PID ${agentPid} got terminated during setup, exiting.`
+    );
+    process.removeAllListeners('disconnect');
+    process.exitCode = 42;
```

Comment:
41 means that the parent got terminated during setup, 42 means the agent got terminated during setup. 😏

I just needed some custom codes so that they don't clash with [other status codes that Node.js uses](https://nodejs.org/docs/latest-v18.x/api/process.html#exit-codes). I'll add a comment to explain they're custom.